### PR TITLE
Add a parameter for custom log4j properties file.

### PR DESCRIPTION
**Pull request by ronin13**
Many a time (such as logging to multiple files with ROLLING appender) it is
a better to pass a custom file (templated higher up in modules calling
this module) to which /etc/zookeeper/conf/log4j.properties is
symlinked.

Also updated the spec tests.

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -36,6 +36,7 @@
   # interval in hours, purging enabled when >= 1
   $purge_interval   = 0,
   # log4j properties
+  $log4j_file   = undef,
```
Comment (solarkennedy) [inline manifests/init.pp]: Can you add a docstring in puppet format to explain how to use this?
```
# == Parameters
#
# [*log4j_file*]
# Needs a full path bla bla
#
```

File: spec/classes/config_spec.rb [hunk 2]
```diff
@@ -77,6 +77,18 @@
     }
   end
 
+  context 'log4j file' do
+    log4j_file = '/tmp/log4j.properties'
+    File.open(log4j_file, "w") { |file| file.write("log4j.rootLogger=INFO, CONSOLE") }
```
Comment (solarkennedy) [inline spec/classes/config_spec.rb]: Whoa here. There is no need to do fileio in the tests.

File: manifests/init.pp (hunk 1 above)
Comment (solarkennedy) [inline manifests/init.pp]: Also please note it is up to the user to get this file deployed and that is outside the scope of this module

**Comment by solarkennedy** [comment]
We'll probably never update to mainline (non-fork), but if we did, would it's log4j stuff be enough for your usecase?
https://github.com/deric/puppet-zookeeper/blob/master/manifests/init.pp#L47

File: manifests/init.pp (hunk 1 above)
Comment (ronin13) [inline manifests/init.pp]: I will add docstring, and yeah, deployment (of target) is outside the scope, so just the link creation is in this module.

File: spec/classes/config_spec.rb (hunk 2 above)
Comment (ronin13) [inline spec/classes/config_spec.rb]: Otherwise I will have to remove that test. Can I just create the empty file instead?

**Comment by ronin13** [comment]
> We'll probably never update to mainline (non-fork), but if we did, would it's log4j stuff be enough for 
> your usecase?
> https://github.com/deric/puppet-zookeeper/blob/master/manifests/init.pp#L47

The log4j bits from that are already in ours, that cannot be used for this use-case.

File: spec/classes/config_spec.rb (hunk 2 above)
Comment (ronin13) [inline spec/classes/config_spec.rb]: Done and pushed.

Comment (solarkennedy) [inline spec/classes/config_spec.rb]: Puppet rspec tests do not actually deploy files, they just inspect what "would" happen by looking at the catalog.

Just inject params and assert that it should contain_file with that line. And maybe also that it should `contain_file(mainlog4jfile).with_link(log4j_file)`

No need to do file io at all, again the tests here just inspect the catalog.

File: manifests/init.pp (hunk 1 above)
Comment (solarkennedy) [inline manifests/init.pp]: perfect

File: spec/classes/config_spec.rb (hunk 2 above)
Comment (ronin13) [inline spec/classes/config_spec.rb]: Thanks, fixed.

File: spec/classes/config_spec.rb [hunk 3]
```diff
@@ -77,6 +77,19 @@
     }
   end
 
+  context 'log4j file' do
+    let(:params) { {
+      :log4j_file    => "/nail/etc/zookeeper/log4j.properties",
+    } }
+
+    it do
+      should contain_file('/etc/zookeeper/conf/zoo.cfg').with(
```
Comment (solarkennedy) [inline spec/classes/config_spec.rb]: I'm surprised this passes, it is not correct

Comment (ronin13) [inline spec/classes/config_spec.rb]: Sorry for that. Fixed it.

Regards,
-- 
Raghavendra "Raghu" Prabhu
GPG Id : 0xD72BE977 | Fingerprint: B93F EBCB 8E05 7039 CD3C A4B8 A616 DCA1
D72B E977
wnohang.net   | about.me/raghavendra.prabhu

**Comment by solarkennedy** [comment]
ship
